libgadget/uvbg.c

```diff
@@ -0,0 +1,1204 @@
+    //TODO: do I need to reinitialise this after petapm? I don't see pfft_cleanup anywhere
+    pfftf_init();
+    pfftf_plan_with_nthreads(omp_get_max_threads());
+
+    //TODO: have the flags stored somewhere so it's not both here and in create_plans
+    unsigned pfft_flags = PFFT_PADDED_R2C|PFFT_TRANSPOSED_NONE;
+
+    // Assign the slab size
+    int n_ranks;
+    MPI_Comm_size(MPI_COMM_WORLD, &n_ranks);
+
+    //TODO(jdavies): check if pfft wants in or out dimensions here
+    ptrdiff_t n[3] = {uvbg_dim,uvbg_dim,uvbg_dim};
+    int np[2];
+
+    /* try to find a square 2d decomposition */
+    //TODO(jdavies): ask Simeon about what this is/does and whether I need it
+    int i;
+    for(i = sqrt(n_ranks) + 1; i >= 0; i --) {
+        if(n_ranks % i == 0) break;
+    }
+    np[0] = i;
+    np[1] = n_ranks / i;
+
+    message(0, "Using 2D Task mesh for UVBG %td x %td \n", np[0], np[1]);
+    if( pfftf_create_procmesh_2d(MPI_COMM_WORLD, np[0], np[1], &UVBGgrids.comm_cart_2d) ){
+        endrun(0, "Error: This test file only works with %td processes.\n", np[0]*np[1]);
+    }
+
+    MPI_Comm comm_cart_2d = UVBGgrids.comm_cart_2d;
+
+    // find out what slab each rank should get
+    ptrdiff_t local_ni[3], local_i_start[3];
+    ptrdiff_t local_no[3], local_o_start[3];
+    ptrdiff_t local_n_complex = pfftf_local_size_dft_r2c_3d(n, comm_cart_2d, pfft_flags, local_ni, local_i_start, local_no, local_o_start);
+
+    // let every rank know...
+    ptrdiff_t* slab_ni = mymalloc("slab_ni",sizeof(ptrdiff_t) * n_ranks * 3); ///< array of number of x cells of every rank
+    UVBGgrids.slab_ni = slab_ni;
+    MPI_Allgather(&local_ni, sizeof(ptrdiff_t)*3, MPI_BYTE, slab_ni, sizeof(ptrdiff_t)*3, MPI_BYTE, MPI_COMM_WORLD);
+
+    ptrdiff_t *slab_i_start = mymalloc("slab_i_start",sizeof(ptrdiff_t) * n_ranks * 3); ///< array first x cell of every rank
+    UVBGgrids.slab_i_start = slab_i_start;
+    MPI_Allgather(&local_i_start, sizeof(ptrdiff_t)*3, MPI_BYTE, slab_i_start, sizeof(ptrdiff_t)*3, MPI_BYTE, MPI_COMM_WORLD);
+
+    ptrdiff_t* slab_no = mymalloc("slab_no",sizeof(ptrdiff_t) * n_ranks * 3); ///< array of number of x cells of every rank
+    UVBGgrids.slab_no = slab_no;
+    MPI_Allgather(&local_no, sizeof(ptrdiff_t)*3, MPI_BYTE, slab_no, sizeof(ptrdiff_t)*3, MPI_BYTE, MPI_COMM_WORLD);
+
+    ptrdiff_t *slab_o_start = mymalloc("slab_o_start",sizeof(ptrdiff_t) * n_ranks * 3); ///< array first x cell of every rank
+    UVBGgrids.slab_o_start = slab_o_start;
+    MPI_Allgather(&local_o_start, sizeof(ptrdiff_t)*3, MPI_BYTE, slab_o_start, sizeof(ptrdiff_t)*3, MPI_BYTE, MPI_COMM_WORLD);
+
+    ptrdiff_t *slab_n_complex = mymalloc("slab_n_complex",sizeof(ptrdiff_t) * n_ranks); ///< array of allocation counts for every rank
+    UVBGgrids.slab_n_complex = slab_n_complex;
+    MPI_Allgather(&local_n_complex, sizeof(ptrdiff_t), MPI_BYTE, slab_n_complex, sizeof(ptrdiff_t), MPI_BYTE, MPI_COMM_WORLD);
```

Comment:
This doesn't look right to me - the allocations will be a little different on different ranks and with different padding. Try to reuse the code in petapm.c, especially the petapm_region_init_strides function. There is also a pm struct which will be useful.

Reply:
These are meant to be arrays of the size and offsets of each region that are stored on all ranks. If I'm reading the init_strides function correctly, this sets the strides to `(size[2]*size[1], size[2], 1)`. The `grid_index` function does the same thing, except it takes `size[1]` as an argument and calculates `size[2]` based on whether the grid is in realspace, is padded for inplace fft, or in k-space (I've checked and these dimensions match the size arrays). I definitely think its a good idea to re-use the petapm functions, but I think the indexing issue is elsewhere. I'm probably missing something so I'll replace the functions and see if I get the same issue

Reply:
I'm not completely sure where, but I definitely remember tripping over a subtlety with the indexing before, which petapm.c takes care of. It might not be this, because what you say is roughly what they do. But I would still ask you to reuse the functions because it is good practice to have things implemented in only one place.

Reply:
I fixed the indexing issue in the filter function (I really should have caught it earlier), and I'm getting reasonable bubbles again.

Reply:
Yes that looks good! Nice.